File: typescript/src/agentchat/terminations/StopMessageTermination.ts
import { ITerminationCondition } from "../abstractions/Termination";
import { AgentMessage, StopMessage } from "../abstractions/Messages";

/**
 * Terminate a conversation if a StopMessage is received.
 */
export class StopMessageTermination implements ITerminationCondition {
    private _isTerminated = false;

    /**
     * Gets whether the chat has already terminated.
     */
    get isTerminated(): boolean {
        return this._isTerminated;
    }

    /**
     * Checks if new messages should cause termination of the chat.
     * @param messages The messages to check
     * @returns A StopMessage if a stop message is found, null otherwise
     */
    async checkAndUpdateAsync(messages: AgentMessage[]): Promise<StopMessage | null> {
        const stopMessage = messages.find(m => m instanceof StopMessage) as StopMessage | undefined;
        if (stopMessage) {
            this._isTerminated = true;
        }
        return stopMessage ?? null;
    }

    /**
     * Resets the termination condition to its initial state.
     */
    reset(): void {
        this._isTerminated = false;
    }
}

File: dotnet/src/Microsoft.AutoGen/AgentChat/Terminations/StopMessageTermination.cs
using Microsoft.AutoGen.AgentChat.Abstractions;

namespace Microsoft.AutoGen.AgentChat.Terminations;

/// <summary>
/// Terminate a conversation if a <see cref="StopMessage"/> is received.
/// </summary>
public class StopMessageTermination : ITerminationCondition
{
    public bool IsTerminated { get; private set; }

    public ValueTask<StopMessage?> CheckAndUpdateAsync(IList<AgentMessage> messages)
    {
        if (this.IsTerminated)
        {
            throw new TerminatedException();
        }

        foreach (AgentMessage item in messages)
        {
            if (item is StopMessage)
            {
                this.IsTerminated = true;

                StopMessage result = new() { Content = "Stop message received", Source = nameof(StopMessageTermination) };
                return ValueTask.FromResult<StopMessage?>(result);
            }
        }

        return ValueTask.FromResult<StopMessage?>(null);
    }

    public void Reset()
    {
        this.IsTerminated = false;
    }
}
